src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyStatus.cs

```diff
@@ -0,0 +1,300 @@
+        internal const int AdjustmentRunWindowSeconds = 10;
+        internal const int DefaultMinAdjustmentFrequencySeconds = 5;
+        internal const int MaxAdjustmentDelta = 5;
+
+        private readonly ConcurrencyManager _concurrencyManager;
+        private readonly object _syncLock = new object();
+
+        private int _adjustmentRunDirection;
+        private int _adjustmentRunCount;
+
+        /// <summary>
+        /// Constructs a new instance.
+        /// </summary>
+        /// <param name="functionId">The function ID this status is for.</param>
+        /// <param name="concurrencyManager">The <see cref="ConcurrencyManager"/>.</param>
+        public ConcurrencyStatus(string functionId, ConcurrencyManager concurrencyManager)
+        {
+            if (string.IsNullOrEmpty(functionId))
+            {
+                throw new ArgumentNullException(nameof(functionId));
+            }
+            if (concurrencyManager == null)
+            {
+                throw new ArgumentNullException(nameof(concurrencyManager));
+            }
+
+            FunctionId = functionId;
+            _concurrencyManager = concurrencyManager;
+            CurrentConcurrency = 1;
+            OutstandingInvocations = 0;
+            LastConcurrencyAdjustmentTimestamp = DateTime.UtcNow;
+            MaxConcurrentExecutionsSinceLastAdjustment = 0;
+            _adjustmentRunDirection = 0;
+        }
+
+        /// <summary>
+        /// Gets the function ID this status is for.
+        /// </summary>
+        public string FunctionId { get; }
+
+        /// <summary>
+        /// Gets the recommended amount of new work this function can safely process.
+        /// When throttling is enabled, this may return 0 meaning no new work should be
+        /// started.
+        /// </summary>
+        public int FetchCount
+        {
+            get
+            {
+                if (_concurrencyManager.ThrottleEnabled || OutstandingInvocations > CurrentConcurrency)
+                {
+                    // we can't take any work right now
+                    return 0;
+                }
+                else
+                {
+                    // no throttles are enabled, so we can take work up to the current concurrency level
+                    return CurrentConcurrency - OutstandingInvocations;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the current concurrency level for this function. This adjusts
+        /// dynamically over time.
+        /// </summary>
+        public int CurrentConcurrency { get; internal set; }
+
+        /// <summary>
+        /// Gets the current number of in progress invocations of this function.
+        /// </summary>
+        public int OutstandingInvocations { get; internal set; }
+
+        /// <summary>
+        /// Gets the recommended delay for the next time <see cref="ConcurrencyManager.GetStatus(string)"/> should be called
+        /// for this function./>
```

Comment:
Initially I didn't have this value at all, and it was up to listeners to determine how long they wanted to delay when GetStatus returned a FetchCount of 0. So there were hardcoded values. I decided that rather than force extensions to think about this or hardcode a value, the recommended pattern would be to just use this value determined by CM. While internally it's still just hardcoded, it allows us in the future to potentially vary this based on internal metrics.